app/models/work.rb

```diff
@@ -436,40 +461,33 @@ def recipients=(recipient_names)
       gift = self.gifts.for_name_or_byline(name).first
       if gift
         gifts << gift # new gifts are added after saving, not now
-        new_recipients << name unless self.posted # all recipients are new if work not posted
+        new_gifts << gift unless self.posted # all gifts are new if work not posted
       else
-        # check that the gift would be valid
-        g = Gift.new(work: self, recipient: name)
+        g = self.gifts.new(work: self, recipient: name)
         if g.valid?
-          new_recipients << name # new gifts are added after saving, not now
+          new_gifts << g # new gifts are added after saving, not now
         else
-          errors.add(:base, ts("You cannot give a gift to the same user twice."))
+          g.errors.full_messages.each { |msg| self.errors.add(:base, msg) }
         end
       end
     end
-    self.new_recipients = new_recipients.uniq.join(",")
     self.gifts = gifts
+    self.new_gifts = new_gifts
   end
 
   def recipients(for_form = false)
     names = (for_form ? self.gifts.not_rejected : self.gifts).collect(&:recipient)
-    unless self.new_recipients.blank?
-      self.new_recipients.split(",").each do |name|
-        names << name unless names.include? name
-      end
-    end
-    names.join(",")
+    names << self.new_gifts.collect(&:recipient) if self.new_gifts.present?
+    names.flatten.uniq.join(",")
   end
 
-  def save_new_recipients
-    unless self.new_recipients.blank?
-      self.new_recipients.split(',').each do |name|
-        gift = self.gifts.for_name_or_byline(name).first
-        unless gift.present?
-          g = Gift.new(recipient: name, work: self)
-          g.save
-        end
-      end
+  def save_new_gifts
+    return if self.new_gifts.blank?
+
+    self.new_gifts.each do |gift|
+      next if self.gifts.for_name_or_byline(gift.recipient).present?
+
+      Gift.create(recipient: gift.recipient, work: self)
```

Comment:
This line works, though it looks a bit odd. Same question as https://github.com/otwcode/otwarchive/pull/3981#discussion_r668239587.

Reply:
I don't remember at all, so I've pushed the broken version (https://github.com/otwcode/otwarchive/pull/4110/commits/88c12e48d16f9201eded3d8ac1cac376270cfac0) so we can see what fails.

Reply:
I think the issue is that `work_id` doesn't get set properly on gifts created before the work is saved (or at least, gifts created with this particular series of steps). Hence re-creating the gift after the work has been saved.